File: src/routing/orsGeocoder.js
/**
 * OpenRouteService geocoding (search-by-name/address) — Pelias-based,
 * https://openrouteservice.org/dev/#/api-docs/geocode/search/get
 *
 * Same CONFIG.ORS_API_KEY already used for routing (src/routing/orsProvider.js)
 * — ORS's free/Standard plan shares one key across Directions, Geocoding,
 * Isochrones, and Matrix under a combined daily quota, so this needed no
 * separate signup/key.
 */
const OrsGeocoder = (() => {
  const BASE_URL   = "https://api.openrouteservice.org/geocode/search";
  const TIMEOUT_MS = 8000;
  const MIN_CHARS  = 3; // shorter queries are mostly noise/wasted quota

  // 2026-09-21: `focus.point` ALONE only biases ranking, it never excludes a
  // far-away match — confirmed directly from Pelias's own real docs (ORS's
  // geocode/search endpoint is a hosted Pelias instance, same param names,
  // confirmed against GIScience/openrouteservice-py's own official client
  // source): "unlike a boundary.circle query, important results far from
  // the given coordinate may still be returned... a query for 'Paris' with
  // a focus.point in Texas [can] return both Paris, TX and Paris, France."
  // Reported directly: searching a real, nearby named business (a chain
  // daycare centre in the user's own home town) surfaced results in
  // completely different countries. Pelias's own docs recommend combining
  // focus.point with boundary.circle.* for exactly this "nearest X within
  // N km" case — CONFINE_RADIUS_KM is a reasonable starting guess for a
  // driving-nav app's realistic day-trip range (generous enough not to
  // exclude a legitimately-searched-for city a few hours away), not tuned
  // against real field data yet, same honest provenance this codebase
  // already carries for its other tuned constants (e.g. visibility.js's
  // CONTRAIL_* thresholds).
  const CONFINE_RADIUS_KM = 200;

  /**
   * 2026-09-21, same day: `boundary.circle` genuinely EXCLUDES anything
   * outside CONFINE_RADIUS_KM, not just deprioritises it — a real, direct
   * consequence of the fix above, reported immediately: "what happens if
   * I'm navigating somewhere further than 200nm away?" A destination
   * beyond ~108nm/200km would have silently returned zero results from
   * this confined query, with nothing upstream (ActiveGeocoder, TomTom's
   * own identical-radius confinement) able to recover it. Fixed by
   * retrying WITHOUT boundary.circle (focus.point bias only, the original
   * pre-fix behaviour) the moment a confined search comes back empty —
   * confined-first still solves the reported "wrong country" bug for any
   * search that has a real nearby match, and only falls back to the wider,
   * less-safe unconfined search for the genuinely rarer case where nothing
   * local exists at all, so a long-distance destination still resolves.
   */
  async function search(text, focus) {
    const query = (text || "").trim();
    if (query.length < MIN_CHARS) return [];

    const apiKey = CONFIG.ORS_API_KEY;
    if (!apiKey) {
      console.warn("OrsGeocoder: CONFIG.ORS_API_KEY not set — see config.js");
      return [];
    }

    if (focus) {
      const confined = await _fetch(query, apiKey, focus, true);
      if (confined.length > 0) return confined;
      return _fetch(query, apiKey, focus, false);
    }
    return _fetch(query, apiKey, null, false);
  }

  async function _fetch(query, apiKey, focus, confine) {
    const params = new URLSearchParams({
      api_key: apiKey,
      text: query,
      size: "6",
    });
    if (focus) {
      params.set("focus.point.lon", focus.lon);
      params.set("focus.point.lat", focus.lat);
      if (confine) {
        params.set("boundary.circle.lon", focus.lon);
        params.set("boundary.circle.lat", focus.lat);
        params.set("boundary.circle.radius", String(CONFINE_RADIUS_KM));
      }
    }

    const controller = new AbortController();
    const timer      = setTimeout(() => controller.abort(), TIMEOUT_MS);

    try {
      const res = await fetch(`${BASE_URL}?${params.toString()}`, { signal: controller.signal });
      clearTimeout(timer);
      if (!res.ok) return [];

      const data = await res.json();
      return (data.features || [])
        .filter(f => f.geometry && Array.isArray(f.geometry.coordinates))
        .map(f => ({
          label: (f.properties && f.properties.label) || query,
          lon: f.geometry.coordinates[0],
          lat: f.geometry.coordinates[1],
        }));
    } catch (err) {
      clearTimeout(timer);
      console.warn("OrsGeocoder: search failed —", err.message);
      return [];
    }
  }

  return { search };
})();

if (typeof module !== "undefined") module.exports = OrsGeocoder;

File: android/app/src/main/java/org/vectair/vcas/car/logic/OrsGeocoder.kt
package org.vectair.vcas.car.logic

import org.json.JSONObject

/**
 * OpenRouteService geocoding (search-by-name/address) — a structural port
 * of `src/routing/orsGeocoder.js`, same split as `OrsProvider.kt`: a pure,
 * testable `parseSearchResponse()` and the real network call. Same
 * Pelias-based GeoJSON response shape the JS original reads:
 * `features[].properties.label` / `features[].geometry.coordinates`
 * (`[lon, lat]`).
 */
object OrsGeocoder {

    data class Result(val label: String, val lat: Double, val lon: Double)

    const val MIN_CHARS = 3 // shorter queries are mostly noise/wasted quota, matches the JS original

    fun searchUrl(baseUrl: String, apiKey: String, text: String, focusLat: Double?, focusLon: Double?): String {
        val params = StringBuilder()
        params.append("api_key=").append(java.net.URLEncoder.encode(apiKey, "UTF-8"))
        params.append("&text=").append(java.net.URLEncoder.encode(text, "UTF-8"))
        params.append("&size=6")
        if (focusLat != null && focusLon != null) {
            params.append("&focus.point.lat=").append(focusLat)
            params.append("&focus.point.lon=").append(focusLon)
        }
        return "$baseUrl?$params"
    }

    fun parseSearchResponse(json: JSONObject, fallbackLabel: String): List<Result> {
        val features = json.optJSONArray("features") ?: return emptyList()
        val results = ArrayList<Result>(features.length())
        for (i in 0 until features.length()) {
            val f = features.optJSONObject(i) ?: continue
            val geometry = f.optJSONObject("geometry") ?: continue
            val coords = geometry.optJSONArray("coordinates") ?: continue
            if (coords.length() < 2) continue
            val properties = f.optJSONObject("properties")
            val label = properties?.optString("label")?.takeIf { it.isNotBlank() } ?: fallbackLabel
            results.add(Result(label = label, lon = coords.optDouble(0), lat = coords.optDouble(1)))
        }
        return results
    }

    /** Real network call — see `OrsProvider.kt`'s own doc comment for the same discipline. */
    fun search(apiKey: String, text: String, focusLat: Double?, focusLon: Double?): List<Result> {
        val query = text.trim()
        if (query.length < MIN_CHARS || apiKey.isBlank()) return emptyList()

        val url = searchUrl(BASE_URL, apiKey, query, focusLat, focusLon)
        var connection: java.net.HttpURLConnection? = null
        return try {
            connection = (java.net.URL(url).openConnection() as java.net.HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                setRequestProperty("Accept", "application/json")
            }
            if (connection.responseCode != java.net.HttpURLConnection.HTTP_OK) return emptyList()
            val body = connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            parseSearchResponse(JSONObject(body), query)
        } catch (e: Exception) {
            emptyList()
        } finally {
            connection?.disconnect()
        }
    }

    private const val BASE_URL = "https://api.openrouteservice.org/geocode/search"
    private const val TIMEOUT_MS = 8000
}
